### ISSUE TITLE
Forwards and Backwards compatibility for exceptions

### ISSUE THREAD

**Issue by Raynos** [uber/thriftrw#47]
Thrift does not have a backwards compatible way of adding Exceptions.

This is a terrible thing. This means we cannot add new exceptions to a method at run time. Adding new exceptions is a fact of live; If we do not add new exceptions people will abuse error frames instead.

We should start a discussion on how to approach this problem.

cc @malandrew @kriskowal @prashantv @abhinav

**Comment by Raynos** [comment]
Suggestion:

## Adding a new keyword; `typedException` to thrift.

A `typedException` will have 2 mandatory fields; `message` and `type` with two mandatory numeric ids `100` and `101`. 

This is an IDL only construct; a `typedException` is still a struct on the wire.

There would be validation around the field names and the field types and the field ids in the IDL parser.

When `TChannelThrift` encounters a thrift exception in a `NOT OK` call response that it does not recognise but has fields with id `100` and `101` it will create an instance of `TypedException` with two fields; `type` and `message`.

This strategy would allow us; by convention to create useful error objects in thrift clients when a server adds a new exception to a method.

**Comment by abhinav** [comment]
Instead of adding new syntax, what if we simply standardize that all exceptions must have a field with ID 255 (or 100) and type `string`. We'll always be able to parse the message even if we don't recognize the exception, and then we can construct a generic `ThriftException(message)`.

**Comment by abhinav** [comment]
One problem with standardization is that all existing IDLs now have to change where they put exception messages. Most people are probably using `1: string message`.

**Comment by kriskowal** [comment]
We could make it mandatory for every exception to have a `1: required string message` and have a catcall exception type on all methods.

I will revisit what I did with #26 (TypedError exception spec) which took the message off the wire entirely, which is clearly not going to fly for unrecognized exceptions. We want the message on the wire for those cases.

**Comment by kriskowal** [comment]
cc @Willyham

**Comment by prashantv** [comment]
Brainstorming some different approaches other than having a required "message":
 - Stringify the error on the server side and send that over as well as the original error
 - Once we allow servers to expose Thrift IDL, client could try to be super smart and retrieve the IDL on a new connection which it could use to decode these unknown structs
 - Use TJSONProtocol or similarly JSON encoding to encode Thrift errors so they can still be decoded as Thrift but can also be stringified easily if the Thrift type is not known.

**Comment by abhinav** [comment]
> Once we allow servers to expose Thrift IDL, client could try to be super smart and retrieve the IDL on a new connection which it could use to decode these unknown structs

Will that cause code to start breaking randomly in production because a dependency changed their IDL in a way where the wire representation is compatible but the code needs to change?

> Use TJSONProtocol or similarly JSON encoding to encode Thrift errors so they can still be decoded as Thrift but can also be stringified easily if the Thrift type is not known.

That sounds more plausible, although we don't need TJSON. We can parse the binary payload as-is even if we don't know which class to instantiate with the parsed information. So we could also stuff that information into a dict (it would contain field IDs, not names) and raise a generic client side exception with that.

**Comment by kriskowal** [comment]
My intuition is to follow @abhinav's idea: have an implicit fallback exception that surfaces the numbered fields instead of the named fields.

**Comment by prashantv** [comment]
Yep, JSON has the advantage of having field names, but if we don't think that's important, then raising an implicit exception with numbered fields sounds fine too. Either way, I want to avoid requiring conventions on top that might break existing users etc.

**Comment by Willyham** [comment]
+1 for the generic exception without field names. Seems intuitive enough if we give that structure itself a good name.

**Comment by kriskowal** [comment]
I believe we have cross-language consensus. Here is a concrete proposal, which I am handing off to @malandrew to implement.

If a Thrift result struct is read off the wire and contains an unrecognized field number, the client should create an UnrecognizedThriftException, read that field off the wire an anonymous object graph, and assign that value to a property of the exception. All result types effectively have an implicit `-1: failure UnrecognizedThriftException`.

**Comment by abhinav** [comment]
Just to clarify: We want this behavior for service method responses only, right? Imposing 'no unrecognized fields' on structs is just breaking backwards compatibility across services.

**Comment by kriskowal** [comment]
@abhinav Yes, we are proposing that, when reading a *result* struct off the wire, if we encounter a field we do not recognize, we will reify it as an UnrecognizedException, and add the anonymous/numbered structure on the wire as a property of the exception.

**Comment by Willyham** [comment]
One further clarification: Does this behaviour only happen when `ok` is `false`?

**Comment by abhinav** [comment]
Most likely, yeah. If the protocol is being followed correctly, the only time a non-0 field in the response union is set is when ok is false. Do we want to check that explicitly?
